7.Multiparty-UICollectionView/7.Multiparty-UICollectionView/MultipartyLayout.swift

```diff
@@ -0,0 +1,159 @@
+//
+//  MultipartyLayout.swift
+//  7.Multiparty-UICollectionView
+//
+//  Created by Roberto Perez Cubero on 17/04/2017.
+//  Copyright © 2017 tokbox. All rights reserved.
+//
+
+import UIKit
+
+extension Int {
+    var isEven: Bool {
+        get {
+            return self % 2 == 0
+        }
+    }
+}
+
+class MultipartyLayout: UICollectionViewLayout {
+    fileprivate var cache = [UICollectionViewLayoutAttributes]()
+    fileprivate var cachedNumberOfViews = 0
+    
+    override func prepare() {
+        guard let views = collectionView?.numberOfItems(inSection: 0)
+            else {
+                cache.removeAll()
+                return
+        }
+        
+        if views != cachedNumberOfViews {
+            cache.removeAll()
+        }
+        
+        if cache.isEmpty {
+            cachedNumberOfViews = views
+            let attribs: [UICollectionViewLayoutAttributes] = {
+                switch views {
+                case 1:
+                    return attributesForPublisherFullScreen()
+                case 2:
+                    return attributesForPublisherAndOneSubscriber()
+                case let x where x > 2 && x.isEven:
+                    return attributesForAllViewsTwoByTwo(withNumberOfViews: x)
+                case let x where x > 2 && !x.isEven:
+                    return attributesForPublisherOnTopAndSubscribersTwoByTwo(withNumberOfViews: x)
+                default:
+                    return []
+                }
+            }()
+            
+            cache.append(contentsOf: attribs)
+        }
+    }
+    
+    fileprivate func attributesForPublisherFullScreen() -> [UICollectionViewLayoutAttributes] {
+        var attribs = [UICollectionViewLayoutAttributes]()
+        let ip = IndexPath(item: 0, section: 0)
+        let attr = UICollectionViewLayoutAttributes(forCellWith: ip)
+        attr.frame = collectionView!.superview!.bounds
+        attribs.append(attr)
+        
+        return attribs
+    }
+    
+    // Will layout publisher view over subscriber view
+    fileprivate func attributesForPublisherAndOneSubscriber() -> [UICollectionViewLayoutAttributes] {
+        var attribs = [UICollectionViewLayoutAttributes]()
+        let height = collectionView!.superview!.bounds.size.height / 2
+        let width = collectionView!.superview!.bounds.size.width
```

Comment:
Do we want to use `?` instead of `!`? although it won't affect a lot, it's just a best practise and sense to give to Swift developers.